### ISSUE TITLE
Confused about how to provide GCP creds in backend-beta

### ISSUE THREAD

**Issue by jenspinney** [ljfranklin/terraform-resource#69]
I'm using the [backend-beta](https://github.com/ljfranklin/terraform-resource/tree/WIP-tf-backends#backend-beta) version of this resource and trying to configure it to use GCS.

The docs give this example for configuring the resource for GCS:
```
resources:
  - name: terraform
    type: terraform
    source:
      backend_type: gcs
      backend_config:
        bucket: mybucket
        prefix: mydir
        region: us-central1
        credentials: {{gcp_credentials_json}}
```

My source configuration follows the same pattern and I'm setting the GCP creds variable to be the JSON key provided by Google Cloud. The error I'm running into comes from the `{` character at the beginning of the JSON block. That character seems to make terraform think it's dealing with HCL, and then terraform fails the HCL parsing.

[The documentation of the GCS backend](https://www.terraform.io/docs/backends/types/gcs.html#configuration-variables) config says that this parameter should be a file path, not the actual JSON. But since the source config is defined in the resource definition, I'm not sure how I would inject a file into the container at that point. Maybe I could modify part of the source at `put` time and have a previous step put the GCP creds file in the right spot, but that seems pretty convoluted.

Since GCS is probably a pretty popular backend and I don't see anyone else talking about this, I assume that I'm missing something here.

My exact source configuration looks like this:
```
resources:
  - name: terraform
    type: terraform
    source:
      backend_type: gcs
      backend_config:
        bucket: ((bosh-lite-storage-bucket))
        prefix: terraform_state_files/
        region: ((bosh-lite-region))
        credentials: ((bosh-lite-gcp-json-key))
```

The error I get looks like this:
`2018/10/13 00:36:38 Failed to create env name: terraform init command failed.
Error: exit status 1
Output: invalid value "credentials={\"type\":\"service_account\",\"project_id\":\"<REDACTED>\",\"private_key_id\":\"<REDACTED>\",\"private_key\":\"<REDACTED>\",\"client_email\":\"<REDACTED>\",\"client_id\":\"<REDACTED\",\"auth_uri\":\"https://accounts.google.com/o/oauth2/auth\",\"token_uri\":\"https://accounts.google.com/o/oauth2/token\",\"auth_provider_x509_cert_url\":\"<REDACTED>\",\"client_x509_cert_url\":\"<REDACTED>\"}" for flag -backend-config: Cannot parse value for variable ("{\"type\":\"service_account\",\"project_id\":\"<REDACTED>\",\"private_key_id\":\"<REDACTED>\",\"private_key\":\"<REDACTED>\",\"client_email\":\"<REDACTED>\",\"client_id\":\"<REDACTED\",\"auth_uri\":\"https://accounts.google.com/o/oauth2/auth\",\"token_uri\":\"https://accounts.google.com/o/oauth2/token\",\"auth_provider_x509_cert_url\":\"<REDACTED>\",\"client_x509_cert_url\":\"<REDACTED>\"}") as valid HCL: At 1:12: illegal char
`

which makes sense given Terraform's logic of when to treat something as HCL: https://github.com/hashicorp/terraform/blob/7b55d1640efda8faead1b548fc20264ddfa15a68/helper/variables/parse.go#L45-L64.

I assume others are using the backend-beta with GCS, so I'm wondering what I'm doing wrong here.

**Comment by ljfranklin** [comment]
Hey Jen! I'm having trouble getting GCP backend to work as well, although my test is failing to run `terraform init`:

```
2018/10/15 20:14:04 terraform init command failed.
	
Error: exit status 1
	
Output: Usage: terraform init [options] [DIR]
```

In my CI I only have coverage around S3. Maybe something changed since I wrote the README? I did confirm that the GCS backend does accept either a filepath or the file contents: https://github.com/hashicorp/terraform/blob/48c1ae62b3e1f5e5bc1c2e9ab3e64c7b4b86a6a1/backend/remote-state/gcs/backend.go#L137-L141. I'll take a closer look sometime in the next day or two.

**Comment by ljfranklin** [comment]
cc @cdutra who I think was having similar issues

**Comment by ljfranklin** [comment]
@jenspinney Give it another try, I released [this commit](https://github.com/ljfranklin/terraform-resource/commit/c04bea3ca873ce3c739ef29aac99fa11189ecc36) that should fix it. Terraform was having problems parsing complex values in the `terraform init -backend-config 'creds={...}'` flag. Instead of doing inline values I wrote all the config values to a JSON file and passed that to `-backend-config=./config.json`. I also added a new pipeline group to test that the GCS backend works, not sure if this was a regression or I missed something when I originally wrote the README.
